Add mount move concepts page, update api docs (#14544)
hashicorp/vault: website/content/docs/concepts/mount-migration.mdx

---
layout: docs
page_title: Mount Migration
description: Move secrets and auth mounts within and across namespaces
---

# Mount Migration

> Vault supports moving secrets and auth mounts, both within an across namespaces. The operation

Reviewer: ```suggestion
Vault supports moving secrets and auth mounts, both within, and across namespaces. The operation
```

> is supported by API endpoints as well as a CLI helper around them.

Reviewer: ```suggestion
is supported by API endpoints and a CLI helper around them.
```

## Why

There can be several reasons for wanting to migrate mounts. On OSS, the use cases could be for 

> renaming mounts in order to align with org standards. 

Reviewer: ```suggestion
renaming mounts to align with org standards. 
```

> On Enterprise, there are a few more potential use cases as namespaces come into the picture. 

Reviewer: ```suggestion
There may be more potential for use cases on Enterprise as namespaces come into the picture. 
```

When migrating from an OSS binary to an Enterprise binary, organizations may want to divide their 

> mounts across several namespaces. Mounts may also be moved across namespaces when there is a reorganization

Reviewer: ```suggestion
mounts across several namespaces. Mounts may also be moved across namespaces when roles and responsibilities are necessary for reorganization.
```

Author: Hmm I think this sentence should stay as is, it is roles and responsibilities i.e some secrets engine owned by the core team is given to the devex team's namespace, that are being reorganized. wdyt?

> of roles and responsibilities.

Reviewer: ```suggestion

```

## Async Behavior

The first thing to note about the `sys/remount` endpoint is that it is an asynchronous endpoint. An invocation

> will start the migration process, and the API will return a migration ID. That ID can in turn be used to monitor

Reviewer: ```suggestion
will start the migration process, and the API will return a migration ID. This ID, in turn, be used to monitor
```

> the status of the migration using the `sys/remount/status` endpoint.

Reviewer: ```suggestion
the migration status using the `sys/remount/status` endpoint.
```

## Namespaces

The `sys/remount` endpoint can move mounts across namespaces with the namespaces supplied as prefixes

> on the path arguments. In this case, namespaces for both the source and destination mounts need to be 

Reviewer: ```suggestion
on the path arguments. In this case, namespaces, for both the source and destination mounts, must be 
```

> either the namespace in which the `/sys/remount` endpoint is called, or the children of that namespace. Let us assume that 

Reviewer: ```suggestion
either the namespace in which the `/sys/remount` endpoint is called or the children of that namespace. Let us assume that 
```

> we have 3 namespaces ns1/, ns1/ns2/ and ns1/ns3. We want to move the mount `my-secret` from ns1/ns2/my-secret to ns1/ns3/my-secret. 

Reviewer: ```suggestion
we have three namespaces ns1/, ns1/ns2/, and ns1/ns3. We want to move the mount `my-secret` from ns1/ns2/my-secret to ns1/ns3/my-secret. 
```

> This can be done in one of two ways. The first way is to invoke `/sys/remount` from the root namespace with the `from` argument as ns1/ns2/my-secret 

Reviewer: ```suggestion
Moving the mount can be accomplished in two ways. The first way is to invoke `/sys/remount` from the root namespace with the `from` argument as ns1/ns2/my-secret 
```

and `to` argument as ns1/ns3/my-secret. The second way is to invoke `/sys/remount` from the namespace ns1/, with `from` as ns2/my-secret and `to` 
as ns3/my-secret.

## Leases

> The remount operation will result in existing leases on a mount being revoked. This is potentially the 

Reviewer: ```suggestion
The remount operation will result in existing leases for a mount being revoked. Please note that remounting may be a consuming process. 
```

Author: I think this should stay as is, or be reworded in a different way? I do want it clear that if the remount is taking long it may be cause of this. How about something like this ?
```
On mounts with a large number of leases, the revocation process and hence remounting as a whole may be a time consuming process
```

Author: ```suggestion
The remount operation will result in existing leases on a mount being revoked. On mounts with a large number of leases, the revocation process and hence remounting as a whole may be a time consuming process.
```

> most time consuming part of the operation. Any dynamic secrets for a secrets mount, and any tokens associated

Reviewer: ```suggestion
Any dynamic secrets for a secrets mount and any tokens associated
```

> with an auth mount, will be revoked during the remount process.

Reviewer: ```suggestion
with an auth mount will be revoked during the remount process.
```

Author: i think either this comma should stay or the one after `secrets mount` should be removed also ?

Reviewer: You are correct--come should remain.

## Configurations

> The remount operation will retain all the configurations of the mount. For example, if you create a readonly role

Reviewer: ```suggestion
The remount operation will retain all the configurations of the mount. For example, if you create a read-only role
```

> on a database mount, or a user on an userpass mount, after remounting those will be available at the new location.

Reviewer: ```suggestion
on a database mount or a user on a userpass mount, those will be available at the new location after mounting.
```

Author: imo this should still be `after remounting`? ( i agree with moving it to the end though )

Reviewer: Yes, remounting. And do keep it at the end of the sentence :)

Author: ```suggestion
on a database mount or a user on a userpass mount, those will be available at the new location after remounting.
```

## Cleanup operations

> There are 2 main cleanup operations that Vault operators will need to keep in mind after performing a remount.

Reviewer: ```suggestion
There are two primary cleanup operations that Vault operators will need to keep in mind after performing a remount.
```

### Entities and aliases

> When an auth mount is moved, any aliases that refer to users on the auth mount could now point to an invalid mount.

Reviewer: ```suggestion
Any aliases that refer to users on the auth mount could now point to an invalid mount when an auth mount is removed. Pointing to an invalid mount may not be the case for every instance;
```

Author: think this should be `moved` or `remounted` not `removed` ?

Author: ```suggestion
Any aliases that refer to users on the auth mount could now point to an invalid mount when an auth mount is moved. Pointing to an invalid mount may not be the case for every instance;
```

> This is not necessarily always true, a remount within a namespace will end in the aliases pointing to a valid mount,

Reviewer: ```suggestion
a remount within a namespace will end in the aliases pointing to a valid mount.
```

but a remount across namespaces will always result in the aliases pointing to an invalid mount. In the latter case,

> the Vault operator should find and remove those aliases from the source namespace, along with creating equivalent aliases

Reviewer: ```suggestion
the Vault operator should find and remove those aliases from the source namespace, and create equivalent aliases
```

> and entities for the new mount in the target namespace. If new entities and aliases aren't created in the target namespace,

Reviewer: ```suggestion
and entities for the new mount in the target namespace. If new entities and aliases aren't made in the target namespace,
```

Vault will dynamically generate them upon login operations.

### Policies

Remounting can also result in policies referring to invalid paths. Any policy that refers to a moved secrets engine,

> would need to be updated to point to the new location of the engine, assuming that the policy is in the target namespace 

Reviewer: ```suggestion
would need to be updated to point to the new location of the engine, assuming that the policy is in the target namespace, 
```

Author: there isn't a change in the suggestion box here, was this an accidental comment or did you maybe miss out on adding the suggestion here ? ( figured i'd check before resolving it )

Reviewer: > there isn't a change in the suggestion box here, was this an accidental comment or did you maybe miss out on adding the suggestion here ? ( figured i'd check before resolving it )

No change here; fat fingers ;)

Author: ah gotcha, kk!

> or in a parent of the target namespace, or updated to remove the original mount path if not. 

Reviewer: ```suggestion
or a parent of the target namespace, or updated to remove the original mount path if not. 
```

> Any policies associated with roles on a moved auth mount, would need to be inspected

Reviewer: ```suggestion
Any policies associated with roles on a moved auth mount need to be inspected
```

to see if they still refer to valid paths after the remount operation. Say in namespace ns1/, a child of the root namespace, a role on the mount approle/
has policies granting access to /my-secret. If the mount is moved to another child of the root namespace, say ns2/,
which does not have a secrets engine called my-secret/, the Vault operator would either create an equivalent my-secret/
mount in ns2/, or update the policy to refer to paths valid from the ns2/ namespace.

## Quotas

> It is possible that the mount being moved has rate limit or lease count quotas associated with it. The remount process

Reviewer: ```suggestion
The mount being moved may have a rate limit. or lease count quotas. The remount process
```

Author: rewording suggestion, wdyt?
```
The mount being moved may have rate limit or lease count quotas
```

Author: ```suggestion
The mount being moved may have rate limit or lease count quotas.. The remount process
```

will update these quotas to refer to the new path of the mount.

## Path Filters

> The remount operation will respect any existing path filters. There are 3 use cases in particular that are being handled

Reviewer: ```suggestion
The remount operation will respect any existing path filters. Three uses cases, in particular, are shown below.
```
